fdbserver/MutationTracking.cpp

```diff
@@ -0,0 +1,100 @@
+ * limitations under the License.
+ */
+
+#include <vector>
+#include "fdbserver/MutationTracking.h"
+#include "fdbserver/LogProtocolMessage.h"
+
+#if defined(FDB_CLEAN_BUILD) && MUTATION_TRACKING_ENABLED
+#error "You cannot use mutation tracking in a clean/release build."
+#endif
+
+StringRef debugKey = LiteralStringRef( "\xff/globals/lastEpochEnd" );
+StringRef debugKey2 = LiteralStringRef( "\xff\xff\xff\xff" );
+
+TraceEvent debugMutationEnabled( const char* context, Version version, MutationRef const& mutation ) {
+	if ((mutation.type == mutation.ClearRange || mutation.type == mutation.DebugKeyRange) &&
+			((mutation.param1<=debugKey && mutation.param2>debugKey) || (mutation.param1<=debugKey2 && mutation.param2>debugKey2))) {
+		return std::move(TraceEvent("MutationTracking").detail("At", context).detail("Version", version).detail("MutationType", typeString[mutation.type]).detail("KeyBegin", mutation.param1).detail("KeyEnd", mutation.param2));
+	} else if (mutation.param1 == debugKey || mutation.param1 == debugKey2) {
+		return std::move(TraceEvent("MutationTracking").detail("At", context).detail("Version", version).detail("MutationType", typeString[mutation.type]).detail("Key", mutation.param1).detail("Value", mutation.param2));
+	} else {
+		return std::move(TraceEvent());
+	}
+}
+
+TraceEvent debugKeyRangeEnabled( const char* context, Version version, KeyRangeRef const& keys ) {
+	if (keys.contains(debugKey) || keys.contains(debugKey2)) {
+		return std::move(debugMutation(context, version, MutationRef(MutationRef::DebugKeyRange, keys.begin, keys.end) ));
+	} else {
+		return std::move(TraceEvent());
+	}
+}
+
+TraceEvent debugMessagesAndTagsEnabled( const char* context, Version version, StringRef commitBlob ) {
+	BinaryReader rdr(commitBlob, AssumeVersion(currentProtocolVersion));
+	while (!rdr.empty()) {
+		if (*(int32_t*)rdr.peekBytes(4) == VERSION_HEADER) {
+			int32_t dummy;
+			rdr >> dummy >> version;
+			continue;
+		}
+		TagsAndMessage msg;
+		msg.loadFromArena(&rdr, nullptr);
+		bool logAdapterMessage = std::any_of(
+				msg.tags.begin(), msg.tags.end(), [](const Tag& t) { return t == txsTag || t.locality == tagLocalityTxs; });
+		StringRef mutationData = msg.getMessageWithoutTags();
+		uint8_t mutationType = *mutationData.begin();
+		if (logAdapterMessage) {
+			// Skip the message, as there will always be an idential non-logAdapterMessage mutation
+			// that we can match against in the same commit.
+		} else if (LogProtocolMessage::startsLogProtocolMessage(mutationType)) {
+			BinaryReader br(mutationData, AssumeVersion(rdr.protocolVersion()));
+			LogProtocolMessage lpm;
+			br >> lpm;
+			rdr.setProtocolVersion(br.protocolVersion());
+		} else {
+			MutationRef m;
+			BinaryReader br(mutationData, AssumeVersion(rdr.protocolVersion()));
+			br >> m;
+			TraceEvent&& event = debugMutation(context, version, m);
+			if (event.isEnabled()) {
+				return std::move(event.detail("MessageTags", msg.tags));
```

Comment:
Is it expected that we only return the first matched mutation in a message?
What if there are more than one mutation valid in the debug key range? Not outputting the existed-but-missed mutation could lead to confusion, IMO.

Reply:
I think the right thing to do here longer term would be to make a TraceEventGroup that has the same API as a trace event, and fowards all calls to the N trace events that it internally contains.  That would have to be later work though.  I'll at least document this for now.